x-pack/plugins/reporting/server/lib/task.ts

```diff
@@ -0,0 +1,288 @@
+    const { report: taskJson } = job;
+    const message = `Failing ${taskJson.jobtype} job ${taskJson.id}`;
+
+    // log the error
+    let docOutput;
+    if (error) {
+      logRunner(this.logger, 'error', message, error);
+      docOutput = this._formatOutput(error);
+    } else {
+      logRunner(this.logger, 'error', message);
+    }
+
+    // update the report in the store
+    const store = this.getStore();
+    const report = await store.findReport(taskJson);
+    const completedTime = moment().toISOString();
+    const doc = {
+      completed_at: completedTime,
+      output: docOutput,
+    };
+    await store.setReportFailed(report, doc);
+  }
+
+  _formatOutput(output: TaskRunResult | Error) {
+    const docOutput = {} as TaskRunResult;
+    const unknownMime = false;
+
+    if (isOutput(output)) {
+      docOutput.content = output.content;
+      docOutput.content_type = output.content_type || unknownMime;
+      docOutput.max_size_reached = output.max_size_reached;
+      docOutput.csv_contains_formulas = output.csv_contains_formulas;
+      docOutput.size = output.size;
+      docOutput.warnings =
+        output.warnings && output.warnings.length > 0 ? output.warnings : undefined;
+    } else {
+      const defaultOutput = null;
+      docOutput.content = output.toString() || defaultOutput;
+      docOutput.content_type = unknownMime;
+    }
+
+    return docOutput;
+  }
+
+  private async _performJob(job: TaskParams<unknown>) {
+    if (!this.taskExecutors) {
+      throw new Error(`Task run function factories have not been called yet!`);
+    }
+
+    const { report, payload } = job;
+    logRunner(this.logger, 'info', `Starting job ${report.jobtype} job ${report.id}`);
+
+    // get the run_task function
+    const runner = this.taskExecutors.get(report.jobtype);
+    if (!runner) {
+      throw new Error(`No defined task runner function for ${report.jobtype}!`);
+    }
+
+    // return the run output
+    const cancellationToken = new CancellationToken(); // FIXME: allow to be cancellable
+    return await runner(report.id, payload, cancellationToken);
+  }
+
+  private async _completeJob(job: TaskParams<unknown>, output: TaskRunResult) {
+    const { report: taskJson } = job;
+    logRunner(this.logger, 'info', `Saving ${taskJson.jobtype} job ${taskJson.id}`);
+    const completedTime = moment().toISOString();
+    const docOutput = this._formatOutput(output);
+
+    const store = this.getStore();
+    const doc = {
+      completed_at: completedTime,
+      output: docOutput,
+    };
+    const report = await store.findReport(taskJson);
+    await store.setReportCompleted(report, doc);
+  }
+
+  public register(taskManager: TaskManagerSetupContract) {
+    taskManager.registerTaskDefinitions({
+      [REPORTING_QUEUE_JOB_TYPE]: {
+        type: REPORTING_QUEUE_JOB_TYPE,
+        title: 'Reporting job queue',
+        createTaskRunner: this.createTaskRunner,
```

Comment:
I think the timeout and maxAttempts has to be set here, but if we can do it at schedule time it would be better. Those values are sourced from config which resolves asynchronously.